ssl/ssl_test.cc

```diff
@@ -8853,6 +8853,353 @@ TEST(SSLTest, ProcessTLS13NewSessionTicket) {
+  ASSERT_TRUE(SSL_CTX_set_min_proto_version(server_ctx.get(), TLS1_3_VERSION));
+  ASSERT_TRUE(SSL_CTX_set_max_proto_version(client_ctx.get(), TLS1_3_VERSION));
+  ASSERT_TRUE(SSL_CTX_set_max_proto_version(server_ctx.get(), TLS1_3_VERSION));
+
+  // Configure trust store for server.
+  SSL_CTX_set_client_CA_list(server_ctx.get(), get_test_cert_store());
+
+  bssl::UniquePtr<SSL> client, server;
+  ASSERT_TRUE(CreateClientAndServer(&client, &server, client_ctx.get(),
+                                    server_ctx.get()));
+
+  // Enable sending pha_ext on client side, server may authenticate at any time
+  // after the initial handshake
+  SSL_set_post_handshake_auth(client.get(), 1);
+
+  // Start the handshake, client first flight and write. Sends the extension
+  EXPECT_TRUE(client.get()->s3->pha_enabled);
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_NONE);
+  SSL_do_handshake(client.get());
+
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_EXT_SENT);
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_NONE);
+
+  // First server flight, process client hello and send Server Hello,
+  // Certificate, CertificateVerify
+  SSL_do_handshake(server.get());
+  // extension should be received, nothing pending
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_EXT_RECEIVED);
+  // should be false so CertificateRequest not sent in initial message
+  EXPECT_FALSE(server.get()->s3->hs->cert_request);
+
+  // Initial handshake not over, should not work
+  EXPECT_TRUE(SSL_verify_client_post_handshake(server.get()) == 0);
+
+  // Second client flight, process server messages and send client Finished
+  SSL_do_handshake(client.get());
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_EXT_SENT);
+
+  // Second server flight, process client Finished
+  SSL_do_handshake(server.get());
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_EXT_RECEIVED);
+
+  // Check initialization of PHA_config struct
+  EXPECT_TRUE(server.get()->s3->pha_config != nullptr);
+
+  // Ensure handshake is completed
+  EXPECT_TRUE(server.get()->s3->initial_handshake_complete);
+  EXPECT_TRUE(client.get()->s3->initial_handshake_complete);
+
+  // Sending application data
+  SSL_write(server.get(), "hello", 5);
+  uint8_t msg[5];
+  SSL_read(client.get(), &msg, 5);
+  EXPECT_EQ(Bytes(msg), Bytes("hello"));
+
+  // Server decides to authenticate client
+  SSL_set_verify(server.get(), SSL_VERIFY_PEER, nullptr);
+  // No pending messages on wire
+  EXPECT_FALSE(server.get()->s3->pending_hs_data);
+  // Puts CertificateRequest on the wire but not flushed yet
+  EXPECT_TRUE(SSL_verify_client_post_handshake(server.get()));
+  // Now pending message on wire
+  EXPECT_TRUE(server.get()->s3->pending_hs_data);
+
+  // Message gets flushed
+  SSL_write(server.get(), nullptr, 0);
+  EXPECT_FALSE(server.get()->s3->pending_hs_data);
+
+  // Message read by client and processed
+  SSL_read(client.get(), nullptr, 0);
```

Comment:
What's the expected success/failure of `SSL_read` here?

Reply:
SSL_read itself doesn't really have a success value, we mostly care about tls13_post_handshake which we can't access but can verify success by checking state being copied.